packages/protocol/contracts/identity/Attestations.sol

```diff
@@ -562,35 +604,29 @@ contract Attestations is
     bytes32 seed = getRandom().getBlockRandomness(
       uint256(unselectedRequest.blockNumber).add(selectIssuersWaitBlocks)
     );
-    uint256 numberValidators = numberValidatorsInCurrentSet();
+    (address[] memory issuers, uint256 issuersLength) = getPossibleAttestationIssuers();
+
+    require(unselectedRequest.attestationsRequested <= issuersLength, "not enough issuers");
 
     uint256 currentIndex = 0;
-    address validator;
     address issuer;
-    IAccounts accounts = getAccounts();
 
     while (currentIndex < unselectedRequest.attestationsRequested) {
       seed = keccak256(abi.encodePacked(seed));
-      validator = validatorSignerAddressFromCurrentSet(uint256(seed) % numberValidators);
-      issuer = accounts.signerToAccount(validator);
-
-      if (!accounts.hasAuthorizedAttestationSigner(issuer)) {
-        continue;
-      }
+      uint256 idx = uint256(seed) % issuersLength;
+      issuer = issuers[idx];
 
       Attestation storage attestation = state.issuedAttestations[issuer];
 
-      // Attestation issuers can only be added if they haven't been already.
-      if (attestation.status != AttestationStatus.None) {
-        continue;
-      }
-
       currentIndex = currentIndex.add(1);
       attestation.status = AttestationStatus.Incomplete;
       attestation.blockNumber = unselectedRequest.blockNumber;
       attestation.attestationRequestFeeToken = unselectedRequest.attestationRequestFeeToken;
       state.selectedIssuers.push(issuer);
 
+      issuersLength = issuersLength.sub(1);
```

Comment:
I don't understand this logic, it just pops it out from the end of the list?

Reply:
Added comment, hopefully that helps.